Cython/Compiler/PyrexTypes.py

```diff
@@ -4499,12 +4477,23 @@ def is_template_type(self):
         return self.template_type is None
 
     def specialize_here(self, pos, env, template_values=None):
-        if not self.is_template_type():
-            error(pos, "'%r' has already been specialized" % self )
-            return error_type
         if len(template_values) != 1:
             error(pos, "'%s' takes exactly one template argument." % self.name)
-        return SpecialPythonTypeConstructor(self.name, template_values[0])
+        # return a copy of the template type with special_python_type_constructor_name as an attribute
+        # so it can be identified, and a resolve function that gets back to
+        # the original type (since types are usually tested with "is")
+        new_type = template_values[0]
+        if self.special_python_type_constructor_name == "typing.ClassVar":
+            # classvar must remain a py_object_type
+            new_type = py_object_type
+        if (self.special_python_type_constructor_name == "typing.Optional" and
+                not new_type.is_pyobject):
+            # optional must be a py_object, but can be a specialized py_object
+            new_type = py_object_type
+        type_copy = copy.copy(new_type)
```

Comment:
Seeing these two lines together makes me shiver. Can't we handle this more like what `typedef` does?

Reply:
I'm trying something closer to typedef - using `__getattr__` to do attribute lookup if needed. But I've also mostly avoided it being needed by doing `type = type.resolve()` once we've used the information in the name.